src/buffer_reader.cpp

```diff
@@ -50,7 +50,6 @@ offset_t BufferReader::offset() const
 
 
 void BufferReader::read(char* dest, offset_t offset, zsize_t size) const {
-  ASSERT(offset.v, <, source.size().v);
   ASSERT(offset+offset_t(size.v), <=, offset_t(source.size().v));
   if (! size ) {
```

Comment:
Maximal protection is achieved when both `ASSERT`s are present. If the first `ASSERT` is removed, buggy code calling `read()` with an out-of-bounds `offset` and a negative `size` value can cheat the second/remaining `ASSERT`. A more proper fix is to use `<=` instead of '<' in the first assert.

Reply:
The size type is unsigned so we cannot have a negative size.

Reply:
In C++ the practical equivalent of negative value for an unsigned type is usually the result of an overflow or an underflow. The simplest example is `unsigned(-1)`.

Reply:
Fix using `<=`